diff --git a/extras/startup-scripts/patroni.service b/extras/startup-scripts/patroni.service
index fb23ad9..99abd05 100644
--- a/extras/startup-scripts/patroni.service
+++ b/extras/startup-scripts/patroni.service
@@ -1,8 +1,14 @@
-# This is an example systemd config file for Patroni
-# You can copy it to "/etc/systemd/system/patroni.service",
+# It's not recommended to modify this file in-place, because it will be
+# overwritten during package upgrades.  If you want to customize, the
+# best way is to create a file "/etc/systemd/system/patroni.service",
+# containing
+#	.include /lib/systemd/system/patroni.service
+#	Environment=PATRONI_CONFIG_LOCATION=...
+# For more info about custom unit files, see
+# http://fedoraproject.org/wiki/Systemd#How_do_I_customize_a_unit_file.2F_add_a_custom_unit_file.3F
 
 [Unit]
-Description=Runners to orchestrate a high-availability PostgreSQL
+Description=PostgreSQL high-availability manager
 After=syslog.target network.target
 
 [Service]
@@ -19,6 +25,11 @@ WorkingDirectory=~
 # Where to send early-startup messages from the server
 # This is normally controlled by the global default set by systemd
 #StandardOutput=syslog
+# Location of Patroni configuration
+Environment=PATRONI_CONFIG_LOCATION=/etc/patroni/patroni.yml
+
+# Disable OOM kill on the postmaster
+OOMScoreAdjust=-1000
 
 # Pre-commands to start watchdog device
 # Uncomment if watchdog is part of your patroni setup
@@ -26,7 +37,7 @@ WorkingDirectory=~
 #ExecStartPre=-/usr/bin/sudo /bin/chown postgres /dev/watchdog
 
 # Start the patroni process
-ExecStart=/bin/patroni /etc/patroni.yml
+ExecStart=/usr/bin/patroni ${PATRONI_CONFIG_LOCATION}
 
 # Send HUP to reload from patroni.yml
 ExecReload=/bin/kill -s HUP $MAINPID
